src/st/st0/2DAC8.c

```diff
@@ -712,112 +670,190 @@ void EntityDraculaFinalForm(Entity* self) {
                 g_GameEngineStep = 0xA;
                 g_MenuStep = 0;
                 return;
             }
-            prim = &g_PrimBuf[primIndex];
             self->primIndex = primIndex;
-            self->ext.prim = prim;
             self->flags |= FLAG_HAS_PRIMS;
-            prim->r0 = prim->g0 = prim->b0 = 0;
-            LOW(prim->r1) = LOW(prim->r0);
-            LOW(prim->r2) = LOW(prim->r0);
-            LOW(prim->r3) = LOW(prim->r0);
+            prim = &g_PrimBuf[primIndex];
+            self->ext.prim = prim;
             prim->x0 = prim->x2 = 0;
             prim->x1 = prim->x3 = 0x100;
             prim->y0 = prim->y1 = 0;
             prim->y2 = prim->y3 = 0x100;
+            PGREY(prim, 0) = 0;
+            LOW(prim->r1) = LOW(prim->r0);
+            LOW(prim->r2) = LOW(prim->r0);
+            LOW(prim->r3) = LOW(prim->r0);
             prim->priority = self->zPriority + 0x10;
             prim->drawMode = DRAW_TRANSP | 0x30;
             self->step_s++;
         case 6:
             prim = self->ext.prim;
-            prim->g0 = prim->b0 = prim->r0 = prim->r0 + 0x18;
+            prim->r0 += 0x18;
+            prim->g0 = prim->b0 = prim->r0;
             LOW(prim->r1) = LOW(prim->r0);
             LOW(prim->r2) = LOW(prim->r0);
             LOW(prim->r3) = LOW(prim->r0);
-            if (prim->r0 >= 0x81) {
+            if (prim->r0 > 0x80) {
                 self->step_s++;
             }
             break;
         case 7:
             break;
         }
-        if (--self->ext.dracFinalForm.timer == 0) {
+
+        if (!--self->ext.dracFinalForm.timer) {
             g_api.PlaySfx(SFX_EXPLODE_B);
             self->ext.dracFinalForm.timer = 8;
-            temp_s2 = AllocEntity(&g_Entities[224], &g_Entities[256]);
-            temp_s1 = &D_80181108[self->ext.dracFinalForm.unk84].x;
-            if (temp_s2 != NULL) {
-                CreateEntityFromEntity(E_EXPLOSION, self, temp_s2);
-                temp_s2->params = 3;
-                temp_s2->zPriority = (self->zPriority + 1);
-                temp_s2->posX.i.hi = (temp_s2->posX.i.hi + *temp_s1++);
-                temp_s2->posY.i.hi = (temp_s2->posY.i.hi + *temp_s1++);
+            entity = AllocEntity(&g_Entities[224], &g_Entities[256]);
+            temp_s1 = &D_80181108[self->ext.dracFinalForm.unk84 * 2];
+            if (entity != NULL) {
+                CreateEntityFromEntity(E_EXPLOSION, self, entity);
+                entity->params = 3;
+                entity->zPriority = (self->zPriority + 1);
+                entity->posX.i.hi += temp_s1[0];
+                entity->posY.i.hi += temp_s1[1];
             }
-            self->ext.dracFinalForm.unk84 =
-                ((self->ext.dracFinalForm.unk84 + 1) & 0xF);
+            self->ext.dracFinalForm.unk84++;
+            self->ext.dracFinalForm.unk84 &= 0xF;
         }
+
         break;
     case 0xFF:
 #include "../pad2_anim_debug.h"
     }
 }
 
+static s16 D_80181150[][2] = {
```

Comment:
Had to move this data down because PSP stores the `pad2_anim_debug.h` debug string in `.data` meaning it got out of order otherwise.